src/devtools/views/Profiler/utils.js

```diff
@@ -72,3 +79,183 @@ export const scale = (
+    version: 1,
+    profilingOperationsByRootID: profilingOperationsByRootIDArray,
+    profilingSnapshotsByRootID: profilingSnapshotsByRootIDArray,
+    rendererID,
+    rootID,
+  };
+
+  return profilingSummaryForExport;
+}
+
+function unflattenMap<TKey, TValue>(
+  flatArrayOfPairs: Array<TKey | TValue>
+): Map<TKey, TValue> {
+  const map = new Map();
+  for (let i = 0; i < flatArrayOfPairs.length; i += 2) {
+    map.set(flatArrayOfPairs[i], flatArrayOfPairs[i + 1]);
+  }
+  // Assuming we've got an array where TKey and TValue are interleaved.
+  return ((map: any): Map<TKey, TValue>);
+}
+
+function parseExportedProfilingDataV1(
+  data: ExportedProfilingData | any
+): ImportedProfilingData {
+  // TODO (profiling) Ensure integrity of all deep structures of ExportedProfilingDataV1 prior to type conversion.
+  if (
+    !(
+      data.version === 1 &&
+      Array.isArray(data.profilingOperationsByRootID) &&
+      Array.isArray(data.profilingSnapshotsByRootID) &&
+      Array.isArray(data.commitDetails) &&
+      (typeof data.interactions === 'object' && data.interactions !== null) &&
+      Array.isArray(data.interactions.interactions) &&
+      typeof data.interactions.rootID === 'number' &&
+      (typeof data.profilingSummary === 'object' &&
+        data.profilingSummary !== null)
+    )
+  ) {
+    throw new Error('Malformed profile data version 1.');
+  }
+  const importedProfileData: ImportedProfilingData = {
+    version: data.version,
+    profilingOperations: new Map(
+      data.profilingOperationsByRootID.map(pair => {
+        if (
+          !(
+            Array.isArray(pair) &&
+            pair.length === 2 &&
+            typeof pair[0] === 'number' &&
+            Array.isArray(pair[1])
+          )
+        ) {
+          throw new Error(
+            'Malformed profile data version 1: profilingOperationsByRootID pair'
+          );
+        }
+        return [
+          pair[0],
+          pair[1].map(array => {
+            if (
+              !(Array.isArray(array) && array.every(x => typeof x === 'number'))
+            ) {
+              throw new Error(
+                'Malformed profile data version 1: profilingOperationsByRootID pair[1] array'
+              );
+            }
+            return Uint32Array.from(array);
+          }),
+        ];
+      })
+    ),
+    profilingSnapshots: new Map(
+      data.profilingSnapshotsByRootID.map(pair => {
+        if (
+          !(
+            Array.isArray(pair) &&
+            pair.length === 2 &&
+            typeof pair[0] === 'number' &&
+            Array.isArray(pair[1]) &&
+            pair[1].every(
+              x =>
+                Array.isArray(x) &&
+                x.length === 2 &&
+                typeof x[0] === 'number' &&
+                Array.isArray(x[1])
+            )
+          )
+        ) {
+          throw new Error(
+            'Malformed profile data version 1: profilingSnapshotsByRootID pair'
```

Comment:
This validation feels really heavy-handed, to the point of being hard to read.

I think I'd rather rely on our unit tests to ensure that exporting and re-importing data results in valid data, rather than having this amount of runtime validation. I think runtime validation should maybe be as simple as verifying the expected version number. (If people manually edit exported profiling data, it will break and that's okay.)

I found a bug while testing this branch (probably exists in master too) where exporting then re-importing fails, and the "_Malformed profile data version 1: profilingSnapshotsByRootID pair_" isn't all that helpful an error message anyway 😅 

This is mostly just an "FYI" type comment. I'm going to think about this a little more tomorrow, when my brain is fresh.

Reply:
The comment itself may be tweaked to say something obscure like "Something is wrong with your profile, we couldn't parse it". I believe runtime validation at I/O boundary is useful in addition to tests, it covers different user scenario. You might not care about it, so it's your call to remove that, as well as stick `: any` in such places where there's an assumption about a type but no formal verification. Because you don't use a type-aware runtime helper library like io-ts that generates and hides this code from you, it may be intimidating to write validations by hand, I understand.

Reply:
It's not a matter of it being intimidating. I just don't think that the validation justifies the added complexity.

Reply:
During my refactoring I've changed (fixed) the type structure, but didn't change the version. It's a one-off case, but in such case, even though the tests will pass as I changed them, too, the users will get obscure errors from internals of our program rather than from the I/O boundary. If we had this runtime validation, we wouldn't allow bad data into the typed program. But again, it's your call to say "let it fail on property access or on rendering, we don't really care".

Reply:
That's _not_ what I said. In the event we change the data structure, we would bump the version number as well. The end result would be a dialog explaining that the user is loading an older, unsupported version of profiling data (rather than e.g. "_Malformed profile data version 1: profilingSnapshotsByRootID pair_")